src/node_v8.cc

```diff
@@ -210,6 +211,181 @@ void SetFlagsFromString(const FunctionCallbackInfo<Value>& args) {
+  SetHeapStatistics(writer, isolate);
+  writer->json_objectend();
+  writer->json_end();
+}
+
+GCProfiler::GCProfiler(Environment* env, Local<Object> object)
+    : BaseObject(env, object),
+      start_time(0),
+      current_gc_type(0),
+      state(GCProfilerState::kInitialized),
+      writer_(out_stream_, false) {
+  MakeWeak();
+}
+
+// This function will be called when
+// 1. StartGCProfile and StopGCProfile are called and
+//    JS land does not keep the object anymore.
+// 2. StartGCProfile is called then the env exits before
+//    StopGCProfile is called.
+GCProfiler::~GCProfiler() {
+  if (state != GCProfiler::GCProfilerState::kInitialized) {
+    env()->isolate()->RemoveGCPrologueCallback(BeforeGCCallback, this);
+    env()->isolate()->RemoveGCEpilogueCallback(AfterGCCallback, this);
+  }
+}
+
+JSONWriter* GCProfiler::writer() {
+  return &writer_;
+}
+
+std::ostringstream* GCProfiler::out_stream() {
+  return &out_stream_;
+}
+
+void GCProfiler::New(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args.IsConstructCall());
+  Environment* env = Environment::GetCurrent(args);
+  new GCProfiler(env, args.This());
+}
+
+void GCProfiler::Start(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  GCProfiler* profiler;
+  ASSIGN_OR_RETURN_UNWRAP(&profiler, args.Holder());
+  if (profiler->state != GCProfiler::GCProfilerState::kInitialized) {
+    return;
+  }
+  profiler->writer()->json_start();
+  profiler->writer()->json_keyvalue("version", 1);
+
+  uv_timeval64_t ts;
+  if (uv_gettimeofday(&ts) == 0) {
+    profiler->writer()->json_keyvalue("startTime",
+                                      ts.tv_sec * 1000 + ts.tv_usec / 1000);
+  } else {
+    profiler->writer()->json_keyvalue("startTime", 0);
+  }
+  profiler->writer()->json_arraystart("statistics");
+  env->isolate()->AddGCPrologueCallback(BeforeGCCallback,
+                                        static_cast<void*>(profiler));
+  env->isolate()->AddGCEpilogueCallback(AfterGCCallback,
+                                        static_cast<void*>(profiler));
+  profiler->state = GCProfiler::GCProfilerState::kStarted;
+}
+
+void GCProfiler::Stop(const FunctionCallbackInfo<v8::Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  GCProfiler* profiler;
+  ASSIGN_OR_RETURN_UNWRAP(&profiler, args.Holder());
+  if (profiler->state != GCProfiler::GCProfilerState::kStarted) {
+    return;
+  }
+  profiler->writer()->json_arrayend();
+  uv_timeval64_t ts;
+  if (uv_gettimeofday(&ts) == 0) {
+    profiler->writer()->json_keyvalue("endTime",
+                                      ts.tv_sec * 1000 + ts.tv_usec / 1000);
+  } else {
+    profiler->writer()->json_keyvalue("endTime", 0);
+  }
+  profiler->writer()->json_end();
+  profiler->state = GCProfiler::GCProfilerState::kStopped;
+  args.GetReturnValue().Set(
+      String::NewFromUtf8(env->isolate(), profiler->out_stream()->str().c_str())
+          .ToLocalChecked());
```

Comment:
Since we are taking this from a `std::string`, we should use its size instead of relying on null-termination:

```suggestion
  auto string = profiler->out_stream()->str();
  args.GetReturnValue().Set(
      String::NewFromUtf8(env->isolate(), string.data(), v8::NewStringType::kNormal, string.size())
          .ToLocalChecked());
```

or 

```suggestion
  args.GetReturnValue().Set(
      ToV8Value({}, profiler->out_stream()->str().c_str(), env->isolate())
          .ToLocalChecked());
```

Reply:
Done, thanks.